crates/blockchain/blockchain.rs

```diff
@@ -132,8 +140,177 @@ impl Blockchain {
+
+        let Some(first_block_header) = blocks.first().map(|e| e.header.clone()) else {
+            return Err((ChainError::Custom("First block not found".into()), None));
+        };
+
+        let Some(mut state_trie) = self
+            .storage
+            .state_trie(first_block_header.parent_hash)
+            .map_err(|e| (e.into(), None))?
+        else {
+            return Err((ChainError::ParentNotFound, None));
+        };
+
+        let mut all_receipts: Vec<(BlockHash, Vec<Receipt>)> = vec![];
+        let mut total_gas_used = 0;
+        let chain_config: ChainConfig = self
+            .storage
+            .get_chain_config()
+            .map_err(|e| (e.into(), None))?;
+
+        let mut vm = Evm::new(
+            self.evm_engine,
+            self.storage.clone(),
+            first_block_header.parent_hash,
+        );
+
+        let mut add_block = |block: &Block, i: usize| -> Result<(), ChainError> {
+            let is_first_block = i == 0;
+            let is_last_block = i == blocks.len() - 1;
+
+            let block_hash = block.header.compute_block_hash();
+
+            let parent_header = if is_first_block {
+                // for the first block, we need to query the store
+                let Ok(parent_header) = find_parent_header(&block.header, &self.storage) else {
+                    // If the parent is not present, we store it as pending.
+                    self.storage.add_pending_block(block.clone())?;
+                    return Err(ChainError::ParentNotFound);
+                };
+                parent_header
+            } else {
+                // for the subsequent ones, the parent is the previous block
+                blocks[i - 1].header.clone()
+            };
+
+            // Validate the block pre-execution
+            validate_block(block, &parent_header, &chain_config)?;
+
+            let BlockExecutionResult {
+                account_updates,
+                receipts,
+                requests,
+            } = vm.execute_block_without_clearing_state(block)?;
+
+            validate_gas_used(&receipts, &block.header)?;
+
+            self.storage
+                .apply_account_updates_to_trie(&account_updates, &mut state_trie)?;
+
+            if should_commit_intermediate_tries || is_last_block {
+                let root_hash = state_trie.hash_no_commit();
+                validate_state_root(&block.header, root_hash)?;
+                // commit to db after validating the root
+                state_trie.hash().map_err(StoreError::Trie)?;
+                validate_receipts_root(&block.header, &receipts)?;
+                validate_requests_hash(&block.header, &chain_config, &requests)?;
+            }
+
+            all_receipts.push((block_hash, receipts));
+            total_gas_used += block.header.gas_used;
+
+            last_valid_hash = block_hash;
+
+            Ok(())
+        };
+
+        let interval = Instant::now();
```

Comment:
why is the interval start here and not at the beginning of the function? Regardless, I don't know if this function should be in charge of this, it is getting big. Can't this be done from outside?